### PR TITLE
Add `--complete` auto completion mode to `sqllogictests`

### PR DESCRIPTION
**Pull request by alamb**
# Which issue does this PR close?

Closes  https://github.com/apache/arrow-datafusion/issues/4570
re https://github.com/apache/arrow-datafusion/issues/4460

# Rationale for this change
TLDR I want to be able to write a script and capture expected output easily

See https://github.com/apache/arrow-datafusion/issues/4570

# What changes are included in this PR?

1. Update sqllogictest = "0.10.0"
1.  add a `--complete` command line flag to sqllogictest
2. add documentation on how to update the files

# Are these changes tested?

not directly -- they are a tool for testing

As an example:

As  an example, I ported two tests from aggregates.rs to aggregate.slt like:

```
# array_agg_zero
query U
SELECT ARRAY_AGG([]);

# array_agg_one
query U
SELECT ARRAY_AGG([1]);
```

Note I didn't put in the expected results

Then I ran in complete mode:

```shell
cargo test --test sqllogictests -- aggregate --complete
```

```
 # array_agg_zero
 query U
 SELECT ARRAY_AGG([]);
----
[]

 # array_agg_one
 query U
 SELECT ARRAY_AGG([1]);
----
[[1]]
```

🎉 

# Are there any user-facing changes?

No

### CODE REVIEW

File: datafusion/core/tests/sqllogictests/src/main.rs [hunk 1]
```diff
@@ -78,78 +78,49 @@ pub async fn main() -> Result<()> {
...
+        let mut runner = sqllogictest::Runner::new(DataFusion { ctx, file_name });
+
+        // run each file using its own new DB
+        //
+        // We could run these tests in parallel eventually if we wanted.
+        if options.complete_mode {
+            info!("Using complete mode to complete {}", path.display());
+            let col_separator = " ";
+            let validator = default_validator;
+            update_test_file(path, runner, col_separator, validator)
```
Comment (alamb) [inline datafusion/core/tests/sqllogictests/src/main.rs:112]: This calls the excellent function `update_test_file` from @xxchan added in https://github.com/risinglightdb/sqllogictest-rs/pull/130 👨‍🍳 👌

File: datafusion/core/tests/sqllogictests/test_files/aggregate.slt [hunk 2]
```diff
@@ -1077,3 +1077,17 @@ b 5
 c 4
 d 4
 e 4
+
+
+
+# array_agg_zero
```
Comment (alamb) [inline datafusion/core/tests/sqllogictests/test_files/aggregate.slt:1083]: Writing these tests was amazing -- I just added the queries and the `--complete` mode added the expected results ❤️

**Comment by alamb** [comment]
cc @xxchan @mvanschellebeeck

**Comment by alamb** [comment]
I am quite pleased that this worked so well :bowtie:

**Comment by xudong963** [comment]
I'll review it today, love the feature!!!

File: datafusion/core/tests/sqllogictests/README.md [hunk 3]
```diff
@@ -21,7 +21,11 @@
 
 This is the Datafusion implementation of [sqllogictest](https://www.sqlite.org/sqllogictest/doc/trunk/about.wiki). We use [sqllogictest-rs](https://github.com/risinglightdb/sqllogictest-rs) as a parser/runner of `.slt` files in `test_files`.
 
-#### Running tests
+#### Running tests: Validation Mode
+
+In this model, `sqllogictests` runs the statements and queries in a `.slt` file, comparing the expected output in the fule to the output produced by that run.
```
Comment (xudong963) [inline datafusion/core/tests/sqllogictests/README.md:26]:
```suggestion
In this model, `sqllogictests` runs the statements and queries in a `.slt` file, comparing the expected output in the file to the output produced by that run.
```

File: datafusion/core/tests/sqllogictests/README.md [hunk 4]
```diff
@@ -40,6 +44,17 @@ Run only the tests in `information_schema.slt`:
...
 
+#### Updating tests: Completion Mode
+
+In test script completion mode, `sqllogictests` reads a prototype script and runs the statements and queries against the database engine. The output is is a full script that is a copy of the prototype script with result inserted.
+
+You can update tests by passing the `--complete` argument.
+
+```shell
+# Update ddl.slt with output from running
+cargo test -p datafusion --test sqllogictests -- ddl
```
Comment (xudong963) [inline datafusion/core/tests/sqllogictests/README.md:55]:
```suggestion
cargo test -p datafusion --test sqllogictests -- ddl --complete
```

**Review by xudong963 — APPROVED** [review]
Except for a few minor flaws, others lgtm, thanks @alamb 

Currently, we only have one dir `test_files` which contains all test files. As more and more tests become available, we can categorize them by adding some new subdirs under test_files.

Btw, maybe we can use `clap` and `WalkDir` in the future to replace `Options` struct.

**Pull request merged**